### PR TITLE
Add save_on_foods_ca spider (188 locations)"

### LINKED ISSUE #3096: Spider saveonfoods is broken
During the global build at 2021-09-29-14-42-48, spider **saveonfoods** failed with **0 features** and **0 errors**.

Here's [the log](https://data.alltheplaces.xyz/runs/2021-09-29-14-42-48/logs/saveonfoods.txt) and [the output](https://data.alltheplaces.xyz/runs/2021-09-29-14-42-48/output/saveonfoods.geojson) ([on a map](https://data.alltheplaces.xyz/map.html?show=https://data.alltheplaces.xyz/runs/2021-09-29-14-42-48/output/saveonfoods.geojson))

### LINKED ISSUE #8050: Save-on-Foods (JS Blob)
----
### Brand name

Save-On-Foods

Western Canadian supermarket chain

### Wikidata ID

Q7427974
https://www.wikidata.org/wiki/Q7427974
https://www.wikidata.org/wiki/Special:EntityData/Q7427974.json

### Store finder url(s)

Official Url(s): http://www.saveonfoods.com
- [x] I have run the below and if anything in generated, added it.
- [x] I have manually checked for a storefinder
- [ ] This is not a manufacturer listing outlets operating under their own brands / this will not create duplicates
`pipenv run scrapy sf --brand-wikidata=Q7427974 http://www.saveonfoods.com`
```
```

### PR DESCRIPTION
**Pull request by ShinyDog2636**
This is my first ever code commit, please advise if changes are needed

Fixes #8050
Fixes #3096

### CODE REVIEW

File: locations/spiders/save_on_foods_ca.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+from scrapy import Spider
+
+from locations.items import Feature
+from locations.pipelines.address_clean_up import clean_address
+
+# SaveOnFoods
+
+
+class SaveOnFoodsCASpider(Spider):
+    name = "Save_On_Foods_ca"
```
Comment (CloCkWeRX) [inline locations/spiders/save_on_foods_ca.py:10]:
```suggestion
    name = "save_on_foods_ca"
```

File: locations/spiders/save_on_foods_ca.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+                [store["addressLine1"], store["addressLine2"], store["addressLine3"]]
+            )
+            item["city"] = store["city"]
+            item["postcode"] = store["postCode"]
+            item["country"] = store["country"]
+
+            item["phone"] = store["phone"]
+            item["name"] = store["name"]
+            item["ref"] = store["retailerStoreId"]
+            # StoreHours provided by the API but not parsed by this scraper; please add
```
Comment (CloCkWeRX) [inline locations/spiders/save_on_foods_ca.py:31]: Don't be afraid to go after the opening hours; though this one is a little non standard there seem to be only a few distinct patterns:
```
"openingHours":"Easter Sunday 07:00 AM-06:00 PM,
"openingHours":"Easter Sunday 07:00 AM-06:00 PM,
"openingHours":"Everyday 07:00 AM-09:00 PM",
"openingHours":"Everyday 07:00 AM-10:00 PM",
"openingHours":"Everyday 07:00 AM-10:00 PM,
"openingHours":"Everyday 07:00 AM-11:00 PM",
"openingHours":"Everyday 08:00 AM-07:00 PM",
"openingHours":"Everyday 08:00 AM-08:00 PM",
"openingHours":"Everyday 08:00 AM-09:00 PM",
"openingHours":"Everyday 08:00 AM-10:00 PM",
"openingHours":"Everyday 08:00 AM-10:00 PM,
"openingHours":"Saturday 07:00 AM-10:00 PM,
"openingHours":"Saturday 08:00 AM-08:00 PM,
"openingHours":"Saturday 08:00 AM-09:00 PM,
"openingHours":"Weekdays 07:00 AM-11:00 PM,
"openingHours":"Weekdays 08:00 AM-06:00 PM,
"openingHours":"Weekdays 08:00 AM-10:00 PM,
"openingHours":"Weekdays 09:00 AM-09:00 PM,
```

To tackle a situation like this; there's the `add_ranges_from_string` method. You might simply want to do some string replacements of "Weekdays" to "Mo-Fr" and "Everyday" to "Mo-Su" before passing it to the method:
https://github.com/alltheplaces/alltheplaces/blob/4cda3aae1f656ff6954acb22e04af96ffa1a12ee/locations/spiders/ccc.py#L45

**Comment by CloCkWeRX** [comment]
Welcome!
That looks pretty good - the opening hours would be a nice to have if you can.

**Comment by CloCkWeRX** [comment]
Fixes #8050

**Comment by CloCkWeRX** [comment]
Is it worth skipping anything where status != "Active"
Though the API appears to only output Active at the moment, may avoid bringing in locations that no longer exist or are planned to open.

Is the one store that is `"type":"Corporate"` a head office or a store? May be worth handling that.

File: locations/spiders/save_on_foods_ca.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+class SaveOnFoodsCASpider(Spider):
+    name = "Save_On_Foods_ca"
+    item_attributes = {"brand": "Save-On-Foods", "brand_wikidata": "Q7427974"}
+    start_urls = ["https://storefrontgateway.saveonfoods.com/api/stores"]
+
+    def parse(self, response):
+        for store in response.json()["items"]:
+
+            item = Feature()
```
Comment (CloCkWeRX) [inline locations/spiders/save_on_foods_ca.py:17]: One other trick that is a bit of a timesaver is `DictParser`.

Usage example:
https://github.com/alltheplaces/alltheplaces/blob/4cda3aae1f656ff6954acb22e04af96ffa1a12ee/locations/spiders/dm.py#L31

It is clever enough to look at a hash and the most common variants of specific keys, ie `postcode`  or `postal_code` and extract them.

Though it makes your spider slightly less exact, it does save a bunch of repeated code.

**Comment by iandees** [comment]
@ShinyDog2636 do you think you'll get to these bits of feedback from @CloCkWeRX? It's not a problem, I just don't want to step on your branch if you think you'll finish it off.

**Comment by ShinyDog2636** [comment]
Thank you, @CloCkWeRX, for the helpful feedback.

@iandees Yes, I am working on making these changes. My time is very limited (and I am new to this) so it may take me a while to complete. As long you are okay leaving this open, I will continue to work on this. I appreciate this opportunity to learn.

**Comment by iandees** [comment]
Excellent. Feel free to ask questions if you have any!

**Comment by CloCkWeRX** [comment]
Fixes #3096

**Pull request closed without merging**